File: Elixir/746.exs
# Min Cost Climbing Stairs
# Medium

defmodule Solution do
    @spec min_cost_climbing_stairs(cost :: [integer]) :: integer
    def min_cost_climbing_stairs([h1, h2 | tl]) do
        [x2, x1 | _] = total_cost(tl, [h2, h1])
        min(x2, x1)
    end

    defp total_cost([], total), do: total
    defp total_cost(cost, [h1, h2 | t]) do
        last = min(h1, h2) + hd(cost)
        total = [last] ++ [h1]
        total_cost(tl(cost), total)
    end
end

File: Python/746.py
# Min Cost Climbing Stairs
# Medium

class Solution:
    def minCostClimbingStairs(self, cost: 'list[int]') -> int:
        
        total = [0] * len(cost)
        total[0], total[1] = cost[0], cost[1]
        for i in range(2, len(cost)):
            total[i] = min(total[i-1], total[i-2]) + cost[i]
        
        return min(total[-1], total[-2])
            